token-repository/src/main/scala/io/renku/tokenrepository/repository/creation/RevokeCandidatesFinder.scala

```diff
@@ -34,33 +35,59 @@ private trait RevokeCandidatesFinder[F[_]] {
 private object RevokeCandidatesFinder {
   def apply[F[_]: Async: GitLabClient]: F[RevokeCandidatesFinder[F]] =
     (ProjectTokenDuePeriod[F](), RenkuAccessTokenName[F]())
-      .mapN(new RevokeCandidatesFinderImpl[F](_, _))
+      .mapN(new RevokeCandidatesFinderImpl[F](PerPage(50), _, _))
 }
 
-private class RevokeCandidatesFinderImpl[F[_]: Async: GitLabClient](tokenDuePeriod: Period,
+private class RevokeCandidatesFinderImpl[F[_]: Async: GitLabClient](pageSize: PerPage,
+                                                                    tokenDuePeriod: Period,
                                                                     renkuTokenName: RenkuAccessTokenName
 ) extends RevokeCandidatesFinder[F] {
 
+  import eu.timepit.refined.api.Refined
   import eu.timepit.refined.auto._
+  import eu.timepit.refined.collection.NonEmpty
+  import fs2.Stream
   import io.circe.Decoder
   import io.circe.Decoder._
+  import io.renku.http.tinytypes.TinyTypeURIEncoder._
   import org.http4s.Status.{Forbidden, NotFound, Ok, Unauthorized}
   import org.http4s.circe.jsonOf
   import org.http4s.implicits._
   import org.http4s.{EntityDecoder, Request, Response, Status}
+  import org.typelevel.ci._
+
+  private val endpointName: String Refined NonEmpty = "project-access-tokens"
 
   override def findTokensToRemove(projectId: projects.GitLabId, accessToken: AccessToken): F[List[AccessTokenId]] =
+    Stream
+      .iterate(1)(_ + 1)
+      .evalMap(fetch(_, projectId, accessToken))
+      .map { case (tokens, maybeNextPage) =>
+        tokens.filter(renkuTokens).filter(dueToRefresh).map(_._1) -> maybeNextPage
+      }
+      .takeThrough { case (_, maybeNextPage) => maybeNextPage.nonEmpty }
+      .map(_._1)
+      .compile
+      .toList
+      .map(_.flatten)
```

Comment:
Since this is only fetched to do something with each token, we could leave it as a stream perhaps? Then we don't need to load everything in memory and can do the action(s) on each chunk within the `Stream` monad.

Reply:
Yes, good point. I'll try to follow your suggestion